internal/testutils/broker.go

```diff
@@ -0,0 +1,326 @@
+package testutils
+
+import (
+	"context"
+	// embed is used to embed the configuration template for the mock broker.
+	_ "embed"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/godbus/dbus/v5"
+	"github.com/godbus/dbus/v5/introspect"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	objectPathFmt = "/com/ubuntu/authd/%s"
+	interfaceFmt  = "com.ubuntu.authd.%s"
+)
+
+//go:embed config/broker-cfg
+var brokerConfigTemplate string
+
+type isAuthorizedCtx struct {
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+}
+
+type brokerOptions struct {
+	brokerName string
+}
+
+// BrokerOption is the function signature used to tweak the broker creation.
+type BrokerOption func(*brokerOptions)
+
+// WithBrokerName overrides the default broker name.
+func WithBrokerName(name string) func(o *brokerOptions) {
+	return func(o *brokerOptions) {
+		o.brokerName = strings.ReplaceAll(name, "/", "_")
+	}
+}
+
+// BrokerMock is a broker mock that should be used in the tests.
+type BrokerMock struct {
+	isAuthorizedCalls   map[string]isAuthorizedCtx
+	isAuthorizedCallsMu sync.Mutex
+}
+
+// BrokerBusMock is the D-Bus object that will answer calls for the broker mock.
+type BrokerBusMock struct {
+	broker BrokerMock
+}
+
+// StartBusBrokerMock starts the D-Bus service and exports it on the system bus.
+// It returns the configuration file path for the exported broker.
+func StartBusBrokerMock(t *testing.T, args ...BrokerOption) string {
+	t.Helper()
+
+	opts := brokerOptions{
+		brokerName: "BrokerMock",
+	}
+	for _, opt := range args {
+		opt(&opts)
+	}
+
+	busObjectPath := fmt.Sprintf(objectPathFmt, opts.brokerName)
+	busInterface := fmt.Sprintf(interfaceFmt, opts.brokerName)
+
+	conn, err := dbus.ConnectSystemBus()
+	require.NoError(t, err, "Setup: could not connect to system bus")
+	t.Cleanup(func() { _ = conn.Close() })
+
+	bus := BrokerBusMock{broker: *NewBrokerMock()}
+	err = conn.Export(&bus, dbus.ObjectPath(busObjectPath), busInterface)
+	require.NoError(t, err, "Setup: could not export mock broker")
+
+	err = conn.Export(introspect.NewIntrospectable(&introspect.Node{
+		Name: busObjectPath,
+		Interfaces: []introspect.Interface{
+			introspect.IntrospectData,
+			{
+				Name:    busInterface,
+				Methods: introspect.Methods(&bus),
+			},
+		},
+	}), dbus.ObjectPath(busObjectPath), introspect.IntrospectData.Name)
+	require.NoError(t, err, "Setup: could not export mock broker introspection")
+
+	reply, err := conn.RequestName(busInterface, dbus.NameFlagDoNotQueue)
+	require.NoError(t, err, "Setup: could not request mock broker name")
+	require.Equal(t, reply, dbus.RequestNameReplyPrimaryOwner, "Setup: mock broker name already taken")
+
+	configPath := writeConfig(t, opts.brokerName)
+
+	t.Cleanup(func() {
+		_, _ = conn.ReleaseName(busInterface)
```

Comment:
I'm not sure. But it does look more like the way it's done in the actual bus and I think it's a healthy thing to clean up things as we created/assigned them instead of leaving something behind for the compiler to handle.